### PR TITLE
Set new default WP_POST_REVISIONS for production environment

### PR DESCRIPTION
**Pull request by slackday**
By default WordPress saves every post revision. Over time this causes the database to grow necessary large. Even when using a few custom fields per post the `wp_postmeta` table can easily get to + 100.000 rows in a year. This causes slow performance when joining the postmeta table.

I always limit the number of revisions by setting `WP_POST_REVISIONS` to a reasonable number in production environments. I think it's a good idea to add this variable to `production.php` config so people don't forget about it. I don't think the default to save every revision is so good and it's better to remove this variable if you want the behaviour than to think about adding it to every project (and clean up the database if you forget:). 

Also, atleast in my experience, the WordPress function to "revert" back to older versions is not used much anyway. At least not beyond a couple revisions.

https://codex.wordpress.org/Revisions

### CODE REVIEW

**Comment by austinpray** [comment]
Thank you for the PR! I have two stages of feedback:

1. We should avoid or seek to limit prod-only environment changes. Changes like this should be dropped in config/application.php. I don't want to give anyone any surprises in prod. I don't _really_ expect the average bedrock user to follow the development closely enough to be able to pick on up subtle changes like "bedrock now only saves 10 post revisions in prod only".
2. I have doubts about the claim of negative performance impact of letting the postmeta table grow unchecked.

More on 2:

I understand that tables growing like crazy sounds like a bad thing inherently. However, in some pedestrian worst-case scenario queries I just ran against a wp_postmeta with 380k rows I don't really see a huge problem. *Can you please post some queries where you are running into perf trouble?*

```sql
select count(meta_id) from wp_postmeta;

+----------------+
| count(meta_id) |
+----------------+
|         380913 |
+----------------+
```

Here's a really dumb query that nobody would ever run: join the post title on every single meta field.

```sql
select meta_key, meta_value, post_title from wp_postmeta inner join wp_posts on wp_posts.id = wp_postmeta.post_id;

380281 rows in set (0.74 sec)
```

Nearly a second of runtime is bad but not _that_ bad for what we just asked for. That query is useless and unrealistic.

Something slightly more useful perhaps for analytics or maintenance purposes: grab all the attached images and attach post modified dates to them and also attach the post title to them

```sql
select meta_key, meta_value, post_title, post_modified from wp_postmeta inner join wp_posts on wp_posts.id = wp_postmeta.post_id where meta_key = '_wp_attached_file';

...
| _wp_attached_file | laptop-mockup.png                                                                                                                          | laptop-mockup                                                                                                                          | 2018-06-22 13:39:34 |
| _wp_attached_file | Featured-33.jpg                                                                                                                            | Featured Chris Cooper CarrotCast                                                                                                       | 2018-06-24 11:01:46 |
| _wp_attached_file | Facebook-32.jpg                                                                                                                            | Facebook Chris Cooper CarrotCast                                                                                                       | 2018-06-24 11:02:43 |
| _wp_attached_file | Screen-Shot-2018-06-24-at-11.24.45.png                                                                                                     | google built adwords campaign                                                                                                          | 2018-06-24 11:56:04 |
| _wp_attached_file | Screen-Shot-2018-06-24-at-12.26.37.png                                                                                                     | Facebook Leads                                                                                                                         | 2018-06-24 12:28:09 |
| _wp_attached_file | Screen-Shot-2018-06-24-at-13.18.33.png                                                                                                     | AdWords keyword spend                                                                                                                  | 2018-06-24 13:19:40 |
+-------------------+--------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------+---------------------+
3946 rows in set (0.04 sec)

```

Much more useful and much quicker.

https://codex.wordpress.org/Database_Description#Table:_wp_postmeta wp_postmeta has indexes on meta_key and post_id so this is expected to be fast. Let's try selecting .png images since meta_value is not indexed?

```sql
select meta_key, meta_value, post_title, post_modified from wp_postmeta inner join wp_posts on wp_posts.id = wp_postmeta.post_id where meta_value like '%.png';

| _wp_attached_file            | 30-day-money-back.png                                                                                                                      | 30-day-money-back                                                                                                                                  | 2018-06-22 12:35:24 |
| _wp_attached_file            | dashboard-imac-mockup.png                                                                                                                  | dashboard-imac-mockup                                                                                                                              | 2018-06-22 13:28:45 |
| _wp_attached_file            | laptop-mockup.png                                                                                                                          | laptop-mockup                                                                                                                                      | 2018-06-22 13:39:34 |
| _wp_attached_file            | Screen-Shot-2018-06-24-at-11.24.45.png                                                                                                     | google built adwords campaign                                                                                                                      | 2018-06-24 11:56:04 |
| _wp_attached_file            | Screen-Shot-2018-06-24-at-12.26.37.png                                                                                                     | Facebook Leads                                                                                                                                     | 2018-06-24 12:28:09 |
| _wp_attached_file            | Screen-Shot-2018-06-24-at-13.18.33.png                                                                                                     | AdWords keyword spend                                                                                                                              | 2018-06-24 13:19:40 |
+------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------+---------------------+
2848 rows in set (0.69 sec)

```

Oof 0.69 seconds is kinda slow. But still this is sort of an unrealistic user-facing query since a user hitting your homepage is not scan the entire wp_post table unless you are exposing some kind of weird dashboard. Let's simulate an archive page by limiting our query to 50 posts or so.

```sql
select meta_key, meta_value, post_title, post_modified from wp_postmeta inner join wp_posts on wp_posts.id = wp_postmeta.post_id where meta_value like '%.png' and wp_posts.id >= 26588 and wp_posts.id <= 26638;

+-------------------+----------------------------------------+-------------------------------+---------------------+
| meta_key          | meta_value                             | post_title                    | post_modified       |
+-------------------+----------------------------------------+-------------------------------+---------------------+
| _wp_attached_file | 30-day-money-back.png                  | 30-day-money-back             | 2018-06-22 12:35:24 |
| _wp_attached_file | dashboard-imac-mockup.png              | dashboard-imac-mockup         | 2018-06-22 13:28:45 |
| _wp_attached_file | laptop-mockup.png                      | laptop-mockup                 | 2018-06-22 13:39:34 |
| _wp_attached_file | Screen-Shot-2018-06-24-at-11.24.45.png | google built adwords campaign | 2018-06-24 11:56:04 |
| _wp_attached_file | Screen-Shot-2018-06-24-at-12.26.37.png | Facebook Leads                | 2018-06-24 12:28:09 |
| _wp_attached_file | Screen-Shot-2018-06-24-at-13.18.33.png | AdWords keyword spend         | 2018-06-24 13:19:40 |
+-------------------+----------------------------------------+-------------------------------+---------------------+
6 rows in set (0.00 sec)

```

Alright that's more like it. This is where we see those indexes come into play. So no perf problem in a realistic scenario where you are fetching and joining info on 50 or so posts.

I guess let's try the other way around? Grab every single post with the featured image id joined to it?

```sql
select wp_posts.id, post_title, meta_value as featured_image from wp_posts left join (select post_id, meta_value from wp_postmeta where meta_key = '_thumbnail_id') as wp_postmeta on wp_posts.id = wp_postmeta.post_id;

14987 rows in set (0.48 sec)
```

Slow, but then again this is a shitty query because no end-user will query every single post. Let's try it emulating an archive page with like 50 posts on it.

```sql
select wp_posts.id, post_title, meta_value as featured_image from wp_posts left join (select post_id, meta_value from wp_postmeta where meta_key = '_thumbnail_id') as wp_postmeta on wp_posts.id = wp_postmeta.post_id where wp_posts.id >= 26588 and wp_posts.id <= 26638;

50 rows in set (0.01 sec)
```

Yep works for me.

So TLDR: I don't see the perf problem of letting postmeta grow like crazy. Please post the scenario you were encountering.

**Comment by slackday** [comment]
I totally understand your concern with setting only production variables and changing the expected behaviour. And I understand it can cause more questions from the community than problems it solves. I will close this PR.

Actually I did a bit more digging since I posted this PR and found the root cause of my issue. The performance issue had more to do with how WordPress builds queries with `wp_postmeta` when having more than one join.

I was fetching calendar events where the start date. If calendar events would span more than one day they could have an end date.

The code looked like this:

```
      $query->set('meta_key', 'acf__calendar--start-date');
      $query->set('orderby', 'meta_value_num');
      $query->set('order', 'ASC');
      $query->set('meta_query', array(
        'relation' => 'OR',
        array(
          'key' => 'acf__calendar--date-end',
          'value' => date('Ymd'),
          'compare' => '>=',
          'type' => 'NUMERIC',
        ),
        array(
          'relation' => 'AND',
          array(
            'relation' => 'OR',
            array(
              'key' => 'acf__calendar--date-end',
              'compare' => 'NOT EXISTS',
            ),
            array(
              'key' => 'acf__calendar--date-end',
              'compare' => '=',
              'value' => '',
            ),
          ),
          array(
            'key' => 'acf__calendar--start-date',
            'value' => date('Ymd'),
            'compare' => '>=',
            'type' => 'NUMERIC',
          ),
        ),
      ));
```

It would produce SQL that looked like this and took ~24sec to run:

```
SELECT SQL_CALC_FOUND_ROWS wp_4_posts.ID 
FROM wp_4_posts  
LEFT JOIN wp_4_postmeta ON ( wp_4_posts.ID = wp_4_postmeta.post_id )  
LEFT JOIN wp_4_postmeta AS mt1 ON ( wp_4_posts.ID = mt1.post_id )  
LEFT JOIN wp_4_postmeta AS mt2 ON (wp_4_posts.ID = mt2.post_id AND mt2.meta_key = 'acf__calendar--date-end' )  
LEFT JOIN wp_4_postmeta AS mt3 ON ( wp_4_posts.ID = mt3.post_id )  
LEFT JOIN wp_4_postmeta AS mt4 ON ( wp_4_posts.ID = mt4.post_id ) 
WHERE 1=1  
AND ( 
  wp_4_postmeta.meta_key = 'acf__calendar--start-date'  
  AND ( 
    ( mt1.meta_key = 'acf__calendar--date-end' AND CAST(mt1.meta_value AS SIGNED) >= '20180709' )  
    OR ( 
      ( mt2.post_id IS NULL  OR ( mt3.meta_key = 'acf__calendar--date-end' AND mt3.meta_value = '' ) )  
      AND ( mt4.meta_key = 'acf__calendar--start-date' AND CAST(mt4.meta_value AS SIGNED) >= '20180709' ) 
    ) 
  ) 
) 
AND wp_4_posts.post_type = 'cpt_calendar' 
AND ((wp_4_posts.post_status = 'publish')) 
GROUP BY wp_4_posts.ID
ORDER BY wp_4_postmeta.meta_value+0 
ASC LIMIT 0, 3;
```

I noticed that the website had a large number of post revisions with post_meta because I had forgotten to limit the number of post revisions.

I thought that cleaning it up would resolve the issue and I thought it would be a good default to limit the revisions so others don't run into this issue.

However cleaning the database only got the query time to ~5sec...

So I also rewrote the code to look like this:

```
      $query->set('meta_key', 'acf__calendar--start-date');
      $query->set('orderby', 'meta_value_num');
      $query->set('order', 'ASC');
      $query->set('meta_query', array(
        'relation' => 'OR',
        array(
          'key' => 'acf__calendar--date-end',
          'value' => date('Ymd'),
          'compare' => '>=',
          'type' => 'NUMERIC',
        ),
        array(
          'key' => 'acf__calendar--start-date',
          'value' => date('Ymd'),
          'compare' => '>=',
          'type' => 'NUMERIC',
        ),
      ));
```

And SQL looks like this now:

```
SELECT SQL_CALC_FOUND_ROWS wp_4_posts.ID
FROM wp_4_posts 
INNER JOIN wp_4_postmeta
ON ( wp_4_posts.ID = wp_4_postmeta.post_id ) 
INNER JOIN wp_4_postmeta AS mt1
ON ( wp_4_posts.ID = mt1.post_id )
WHERE 1=1 
AND ( wp_4_postmeta.meta_key = 'acf__calendar--start-date' 
AND ( ( mt1.meta_key = 'acf__calendar--date-end'
AND CAST(mt1.meta_value AS SIGNED) >= '20180709' ) 
OR ( mt1.meta_key = 'acf__calendar--start-date'
AND CAST(mt1.meta_value AS SIGNED) >= '20180709' ) ) )
AND wp_4_posts.post_type = 'cpt_calendar'
AND ((wp_4_posts.post_status = 'publish'))
GROUP BY wp_4_posts.ID
ORDER BY wp_4_postmeta.meta_value+0 ASC
LIMIT 0, 3
```

Which runs in 0,0052s. So yeah... more of a WP issue I guess 😄 

Thanks for your detailed feedback! 👍

**Comment by JulienMelissas** [comment]
@slackday - nice! Thanks for providing some detailed stuff here too :)

We're always looking for some more blog posts, and I do think that limiting post revisions is a handy thing to know. Would you be interested in fleshing all of this out a bit more and writing a post about limiting post revisions? It can be quick, but real-life examples of performance impact are always welcome. Let me know what you think!

**Comment by slackday** [comment]
Cool. I've never written a blog post so that's a nice offer for something to try. I'm interested. I just have to do some more research and find the right angle on this issue.

I'm on vacation right now hence the late response. I'll get back to you on Slack or Discourse when I'm back at work.
